pageserver/src/tenant/timeline.rs

```diff
@@ -1221,59 +1231,85 @@ impl Timeline {
+                    error!("expecting initial size computation task to be in state {Running:?}, got {state:?}")
+                } else {
+                    *state = new_state;
+                }
+                res
+            },
+        );
+    }
+
+    #[instrument(skip_all, fields(tenant = %self.tenant_id, timeline = %self.timeline_id))]
+    async fn initial_logical_size_calculation_task(
+        self: &Arc<Self>,
+        init_lsn: Lsn,
+    ) -> anyhow::Result<()> {
+        let mut timeline_state_updates = self.subscribe_for_state_updates();
+        let self_calculation = Arc::clone(self);
+        tokio::select! {
+            calculation_result = spawn_blocking(move || self_calculation.calculate_logical_size(init_lsn)) => {
+                let calculated_size = calculation_result
+                    .context("Failed to spawn calculation result task")?
+                    .context("Failed to calculate logical size")?;
+                match self.current_logical_size.initial_logical_size.set(calculated_size) {
+                    Ok(()) => info!("Successfully calculated initial logical size"),
+                    Err(existing_size) => {
+                        // This shouldn't happen because we use self.initial_size_computation_running to ensure exlusivity here.
+                        // But if it happens, just complain & report success so there are no further retries.
+                        error!("Tried to update initial timeline size value to {calculated_size}, but the size was already set to {existing_size}, not changing")
+                    }
+                }
+                Ok(())
+            },
+            new_event = async {
+                loop {
+                    match timeline_state_updates.changed().await {
+                        Ok(()) => {
+                            let new_state = *timeline_state_updates.borrow();
+                            match new_state {
+                                // we're running this job for active timelines only
+                                TimelineState::Active => continue,
+                                TimelineState::Broken | TimelineState::Stopping | TimelineState::Suspended => return Some(new_state),
                             }
-                            Ok(())
-                        },
+                        }
+                        Err(_sender_dropped_error) => return None,
                     }
-                }.instrument(info_span!("initial_logical_size_calculation", tenant = %self.tenant_id, timeline = %self.timeline_id)),
-            );
+                }
+            } => {
+                match new_event {
+                    Some(new_state) => info!("Timeline became inactive (new state: {new_state:?}), dropping current connections until it reactivates"),
+                    None => info!("Timeline dropped state updates sender, stopping init size calculation"),
```

Comment:
This was just reindented by this PR, but these message are weird. We're not dropping any connections here. 
```suggestion
                    Some(new_state) => info!("Aborting initial size calculation because timeline became inactive (new state: {new_state:?})"),
                    None => error!("Aborting initial size calculation because state watch was dropped"), // can't happen, the sender is not dropped as long as the Timeline exists
```

Reply:
And I think these should be `bail!`. Othewise we return Ok(()), and mark the state as `Success`, even though we haven't set the size correctly.

Reply:
Agree with your suggested edit and that the error case should be bail, will implement this.

Reply:
Oh, yeah, both cases should be bail.